samples/js/photos/src/photo-grid-item.js

```diff
@@ -43,53 +51,95 @@ export default class PhotoGridItem extends React.Component<void, Props, State> {
   constructor(props: Props) {
     super(props);
     this.state = {
-      size: props.gridSize,
-      sizeIsBest: false,
-      url: props.url,
+      blob: this._getBlob(props),
+      url: null,
     };
     this._parentTop = 0;
     this._parentLeft = 0;
-    this._shouldTransition = true;
+    this._load(this.state.blob);
   }
 
-  componentWillUpdate() {
+  componentWillReceiveProps(nextProps: Props) {
+    const nextBlob = this._getBlob(nextProps);
+    if (nextBlob === this.state.blob) {
+      return;
+    }
+    this.setState({
+      blob: nextBlob,
+    });
+    this._load(nextBlob);
+  }
+
+  componentWillUpdate(nextProps: Props, nextState: State) {
+    if (this.state.url !== nextState.url) {
+      this._releaseBlobURL();
+    }
+
     const rect = ReactDOM.findDOMNode(this).parentElement.getBoundingClientRect();
     this._parentTop = rect.top;
     this._parentLeft = rect.left;
   }
 
+  _getBlob(props: Props): Ref<NomsBlob> {
+    const [w, h] = props.fullscreen ?
+      [props.viewport.clientWidth, props.viewport.clientHeight] :
+      [props.gridSize.width, props.gridSize.height];
+    const [_, blob] = props.photo.getBestSize(w, h);
+    return blob;
+  }
+
+  async _load(blob: Ref<NomsBlob>) {
+    const b = await blob.targetValue(this.props.db);
+    const r = b.getReader();
+    const parts = [];
+    for (;;) {
+      const n = await r.read();
+
+      if (n.done) {
+        break;
+      }
+
+      // might have changed
```

Comment:
It also seems possible (or I'm missing it) that the component may have just unloaded, and you should check for that.